### PR TITLE
Third tranche of NdArray refinement.

### PR DESCRIPTION
**Pull request by bkmartinjr**
Continuation of work on #330. This PR primarily focused on the read/write interface of NdArray

Changes:
* Added types for various paramater constants (eg, read batch-format)
* Ported SOMAMeasurement to the new SOMACollection MutableMapping ABC interface
* alignment of NdArray read/write interface to latest spec and added unit tests.
    * Dense: now centered around dense subarray read/write using Arrow Tensor (and numpy ndarray)
    * Sparse: support both Arrow sparse tensors, Arrow Table, and pandas (read only)
    * NdArray can now read/write more than 2D (the previous implementation was hard-wired to 2 dimensions)
    * Changed the read/write API names to be explicit about what types they consume/produce (eg, `read_tensor`)

### CODE REVIEW

File: apis/python/src/tiledbsoma/soma_dense_nd_array.py [hunk 1]
```diff
@@ -153,148 +146,72 @@ def is_sparse(self) -> Literal[False]:
...
 
+    def write_numpy(self, coords: SOMADenseNdCoordinates, values: np.ndarray) -> None:
+        """ "
+        Write a numpy ``ndarray`` to the user specified coordinates
+        """
+        self.write_tensor(coords, pa.Tensor.from_numpy(values))
+
     # ----------------------------------------------------------------
+    #
+    # TODO: this code seems obsolete given the current API. Can we port the `io` package
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/soma_dense_nd_array.py:211]: would you like to file a tracking issue in GH or shall i?

**Review by johnkerl — APPROVED** [review]

File: apis/python/src/tiledbsoma/io.py [hunk 2]
```diff
@@ -304,20 +311,18 @@ def to_anndata(
         for key in measurement.obsm.keys():
             shape = measurement.obsm[key].shape
             assert len(shape) == 2
-            ncols = shape[1]
-            mat = measurement.obsm[key].read_as_pandas_all()
+            mat = measurement.obsm[key].read_numpy((slice(None),) * len(shape))
             print("OBSM", key, mat.shape)
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/io.py:315]: I'll make a pass in this PR just to expedite.  I was trying to minimize changes outside of NdArray, but this seems totally reasonable to do.

File: apis/python/src/tiledbsoma/soma_dense_nd_array.py (hunk 1 above)
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_dense_nd_array.py:212]: I'll create one (see the additional items above that are "similar")

Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_dense_nd_array.py:212]: #361

File: apis/python/src/tiledbsoma/io.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io.py]: > there is similar debugging code in  

not really similar

a few left-over print statements that should not have survived a single commit should go away

intentionally written tooling that was absolutely essential for me to maintain sanity and operability especially during our initial build-out, needs a home that people can agree on

i'll file an issue to move these to a place we can agree on

[the remark this answers is not in the document]
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io.py]: > If you agree, can we get an issue filed to capture the task?

here it sounds like asking me to file a task; below, you said you would

i'm happy either way

my proposal:

* you write up a task with things you would like re-homed
* i'll comment/discuss there

sound ok?

**Comment by johnkerl** [comment]
Awesome, thanks @bkmartinjr ! :)

**Pull request merged**